Add docs for Denoise Layer plugin (costmap_2d) (#209)
ros-planning/navigation.ros.org: tutorials/docs/filtering_of_noise-induced_obstacles.rst

Requirements
============

It is assumed that ROS 2, Gazebo and TurtleBot3 packages are installed or built locally. Please make sure that Nav2 project is also built locally as it was made in :ref:`build-instructions`.

Tutorial Steps
==============

1. Enable Denoise Layer
-----------------------

Denoise Layer is Costmap2D plugin. You can enable the ``DenoiseLayer`` plugin in Costmap2D by adding ``denoise_layer`` to the ``plugins`` parameter in ``nav2_params.yaml``. You can place it in the ``global_costmap`` and (or) ``local_costmap`` to filter noise on a global or local map. The DenoiseLayer plugin should have the following parameter defined:

- ``plugin``: type of plugin. In our case ``nav2_costmap_2d::DenoiseLayer``.

Full list of parameters supported by ``DenoiseLayer`` are listed at :ref:`denoise` page.

It is important to note that ``DenoiseLayer`` typically should be placed before inflation_layer.
This is required to prevent inflation from noise-induced obstacles.
Moreover, ``DenoiseLayer`` expects a binary costmap with values ``0`` (free cell) or ``255`` (obstacle cell).

> Values less than 255 will be interpreted as free cells when processed (but will not be distorted in the cost map).

Reviewer: From `DenoiseLayer::isBackground()` it seems that we have to treat `INSCRIBED_INFLATED_OBSTACLE (253)` and `LETHAL_OBSTACLE (254)` as obstacles, and sometimes `NO_INFORMATION (255)`. This means that values less that 253 will be interpreted as free cells when processed. Please correct the description accordingly.

Author: https://github.com/ros-planning/navigation.ros.org/pull/209/commits/a2cd0470f9110c45264c9ec25f7d0ce03ffedc17 I suppose there is no need to explain here when ``NO_INFORMATION`` is an obstacle, because it is a small detail?

Reviewer: > I suppose there is no need to explain here when NO_INFORMATION is an obstacle, because it is a small detail?

I also think so.

To enable ``DenoiseLayer`` for both global and local costmaps, use the following configuration:

.. code-block:: text

  global_costmap:
    global_costmap:
      ros__parameters:
        ...
        plugins: ["static_layer", "obstacle_layer", "denoise_layer", "inflation_layer"]
        ...
        denoise_layer:
          plugin: "nav2_costmap_2d::DenoiseLayer"
          enabled: True
  ...
  local_costmap:
    local_costmap:
      ros__parameters:
        ...
        plugins: ["voxel_layer", "denoise_layer", inflation_layer"]
        ...
        keepout_filter:
          plugin: "nav2_costmap_2d::DenoiseLayer"
          enabled: True

.. note::

  The key to success in filtering noise is to understand its type and choose the right ``DenoiseLayer`` parameters.
  The default parameters are focused on fast removal of standalone obstacles.
  More formally, an obstacle is discarded if there are no obstacles among the adjacent eight cells.
  This should be sufficient in typical cases.

>   If some sensor generates intercorrellated noise-induced obstacles and small obstacles in the world are unlikely, small groups of obstacles can be removed.

Author: Fixed

.. _filtering_of_noise-induced_obstacles:

> Filtering of noise-induced obstacles

Reviewer 2: ```suggestion
Filtering of Noise-Induced Obstacles
```

************************************

- `Overview`_
- `Requirements`_
- `Tutorial Steps`_
- `How it works`_

.. image:: images/Filtering_of_noise-induced_obstacles/title.png
    :width: 1000px

Overview
========

> Incorrect definition of the robot position on the map or noisy sensor measurements can cause to errors in ``Voxel Layer`` or ``Obstacle Layer``. As a result, salt and pepper noise may appear on the costmap. This noise creates false obstacles that prevent the robot from finding the best path on the map.

Reviewer 2: ```suggestion
Noisy sensor measurements can cause to errors in ``Voxel Layer`` or ``Obstacle Layer``. As a result, salt and pepper noise may appear on the costmap. This noise creates false obstacles that prevent the robot from finding the best path on the map. While the images above show both salt and pepper noise as well as error due to mislocalization, this layer will only remove sensor noise, not mislocalized artifacts misaligned with the static map.
```

This tutorial shows how to configure filtering of false obstacles caused by noise. This functionality is provided by the ``DenoiseLayer`` costmap layer plugin which will be enabled and used in this document.

Requirements
============

It is assumed that ROS 2, Gazebo and TurtleBot3 packages are installed or built locally. Please make sure that Nav2 project is also built locally as it was made in :ref:`build-instructions`.

Tutorial Steps
==============

1. Enable Denoise Layer
-----------------------

Denoise Layer is Costmap2D plugin. You can enable the ``DenoiseLayer`` plugin in Costmap2D by adding ``denoise_layer`` to the ``plugins`` parameter in ``nav2_params.yaml``. You can place it in the ``global_costmap`` and (or) ``local_costmap`` to filter noise on a global or local map. The DenoiseLayer plugin should have the following parameter defined:

- ``plugin``: type of plugin. In our case ``nav2_costmap_2d::DenoiseLayer``.

Full list of parameters supported by ``DenoiseLayer`` are listed at :ref:`denoise` page.

> It is important to note that ``DenoiseLayer`` typically should be placed before inflation_layer.

Reviewer 2: ```suggestion
It is important to note that ``DenoiseLayer`` typically should be placed before the inflation layer.
```

This is required to prevent inflation from noise-induced obstacles.

> Moreover, ``DenoiseLayer`` treated costmap as a binary map.

Reviewer 2: ```suggestion
Moreover, ``DenoiseLayer`` processes only obstacle information in the costmap.
```

.. code-block:: text

  global_costmap:
    global_costmap:
      ros__parameters:
        ...
        plugins: ["static_layer", "obstacle_layer", "denoise_layer", "inflation_layer"]
        ...
        denoise_layer:
          plugin: "nav2_costmap_2d::DenoiseLayer"
          enabled: True
  ...
  local_costmap:
    local_costmap:
      ros__parameters:
        ...
        plugins: ["voxel_layer", "denoise_layer", inflation_layer"]
        ...
        keepout_filter:
          plugin: "nav2_costmap_2d::DenoiseLayer"
          enabled: True

.. note::

  The key to success in filtering noise is to understand its type and choose the right ``DenoiseLayer`` parameters.
  The default parameters are focused on fast removal of standalone obstacles.
  More formally, an obstacle is discarded if there are no obstacles among the adjacent eight cells.
  This should be sufficient in typical cases.

  If some sensor generates intercorrelated noise-induced obstacles and small obstacles in the world are unlikely, small groups of obstacles can be removed.
  To configure the ``DenoiseLayer`` to such cases and understand how it works, refer to the section `How it works`_.
.. warning::

>   Use this plugin to filter the global costmap with caution. Firstly, it introduces performance-related risks.

Reviewer 2: ```suggestion
  Use this plugin to filter the global costmap with caution. Firstly, it introduces potential performance issues.
```

  For example in case of typically-high-range lidars (20+ meters) update window can be massive making processing time unacceptably long.

>   Secondly, if salt and pepper noise is present in the local costmap, it disappears and is not remembered when going back into that region

Reviewer 2: I don't think this secondly sentence is necessary, remove L81-83.

Author: 3cd2571

  since the local costmap doesn't store data in areas after it has left.
  The global costmap does though, so if we leave an area and the salt/pepper noise was being handled by this later, it would still be there when we came back.

>   This feature can’t be considered problem, but is worth be taken into account.

Reviewer 2: ```suggestion
  It is worth taking this into account as an application designer.
```

It is important to note that ``DenoiseLayer`` typically should be placed before inflation_layer.
This is required to prevent inflation from noise-induced obstacles.
Moreover, ``DenoiseLayer`` treated costmap as a binary map.
Values ``INSCRIBED_INFLATED_OBSTACLE``, ``LETHAL_OBSTACLE`` and optionally ``NO_INFORMATION``

> will be interpreted as obstacle cell. Cells with any other values will be interpreted as ``FREE_SPACE`` when processed.

Reviewer: ```
Cells with any other values will be interpreted as ``FREE_SPACE`` when processed.
->
Cells with any other values will be interpreted as ``FREE_SPACE`` when processed (won't be distorted in the cost map).
```
I think, this is important clarificaiton

Author: https://github.com/ros-planning/navigation.ros.org/pull/209/commits/4ac0320c382fdf14f61ddf835075199261911bf4
